### PR TITLE
[Main] Support device-init grouped linear module too with when not using TE opfuser

### PR DESCRIPTION
**Pull request by zhongbozhu**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Proper single weight support, discrete weight is not affected. Adds many numerical unit tests running on GB200. 

Unit testing:

Note: some tests can only run with TE updated (https://github.com/NVIDIA/TransformerEngine/pull/3224), for CI image, it should just pass without causing errors. 

```
pytest -s -v --confcutdir=tests/unit_tests/transformer/moe tests/unit_tests/transformer/moe/test_grouped_mlp.py

torchrun --nproc_per_node=4 --log-dir /tmp/mcore-single-weight-ut --tee 0:3 --redirects 3 -m pytest -s -v --confcutdir=tests/unit_tests/transformer/moe tests/unit_tests/transformer/moe/test_moe_single_grouped_weight_numerics.py

torchrun --nproc_per_node=4 --log-dir /tmp/mcore-grouped-dispatcher-ut --tee 0:3 --redirects 3 -m pytest -s -v --confcutdir=tests/unit_tests/transformer/moe tests/unit_tests/transformer/moe/test_grouped_tensor_dispatcher_numerics.py
```

E2E testing:

Model - Qwen3.5 VL SFT 
Dispatcher - HybridEP
Experiements - bf16 cublas grouped gemm [single weight ON/OFF] vs. cuteDSL with TE op fuser mxfp8 [single weight ON/OFF] vs mxfp8 cublas grouped gemm [single weight ON/OFF]

image

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaox12 — APPROVED** [review]

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -301,25 +301,74 @@ def __init__(
...
+            align_size = (
+                get_align_size_for_quantization(self.config) if self._use_grouped_tensor else None
+            )
             self.quantization_padding = Fp8Padding(self.num_local_experts, align_size=align_size)
             self.quantization_unpadding = Fp8Unpadding(
                 self.num_local_experts, align_size=align_size
             )
 
+    @staticmethod
+    def _apply_packed_bias(intermediate_parallel, packed_bias, tokens_per_expert, permuted_probs):
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py:316]: Note: only models with routed expert bias will use this

File: megatron/core/transformer/moe/experts.py [hunk 2]
```diff
@@ -634,9 +683,21 @@ def _fused_forward(
 
         # Apply padding if needed
         unpadded_tokens_per_expert = None
+        # Some dispatchers have already padded each expert's token segment before the tokens
+        # reach this module:
+        #   * router padding changes the routing map before dispatch;
+        #   * HybridEP pads as part of its fused dispatch/permute operation;
```
Comment (YangFei1990) [inline megatron/core/transformer/moe/experts.py]: nit: HybridEP/NCCL EP

File: megatron/core/transformer/moe/experts.py [hunk 3]
```diff
@@ -766,11 +842,23 @@ def forward(
 
         # Apply padding if needed
         unpadded_tokens_per_expert = None
-        tokens_per_expert: list[int] = tokens_per_expert.tolist()
         permuted_probs = permuted_probs.unsqueeze(-1)
+        # The token buffer may already contain per-expert padding when padding was performed
+        # before expert compute:
+        #   * router padding modified the routing map before dispatch;
+        #   * HybridEP fused padding into dispatch/permute;
```
Comment (YangFei1990) [inline megatron/core/transformer/moe/experts.py]: nit: same + NCCL EP

File: tests/unit_tests/transformer/moe/test_grouped_tensor_dispatcher_numerics.py [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+
+    # The public MoE contract is unchanged by internal alignment. Run backward as a final check
+    # that padding/unpadding preserved a connected, finite autograd path to the original input.
+    assert output.shape == hidden_states.shape
+    output.float().square().mean().backward()
+    assert hidden_states.grad is not None
+    assert torch.isfinite(hidden_states.grad).all()
+
+
+class TestGroupedTensorDispatcherNumerics:
```
Comment (YangFei1990) [inline tests/unit_tests/transformer/moe/test_grouped_tensor_dispatcher_numerics.py:497]: Could we also add test cases for NCCL EP?

**Review by YangFei1990 — COMMENTED** [review]
Just for my understanding, mcore will pass `moe_use_grouped_tensor `/ `use_transformer_engine_op_fuser` into TE, but it is TE's responsibility to pick the groupgemm backend, which is depending on hardware/dtype/envs, so it might not necessarily follow what exactly provided by mcore, is that right?

**Review by yaox12 — APPROVED** [review]

**Comment by zhongbozhu** [comment]
> Just for my understanding, mcore will pass `moe_use_grouped_tensor `/ `use_transformer_engine_op_fuser` into TE, but it is TE's responsibility to pick the groupgemm backend, which is depending on hardware/dtype/envs, so it might not necessarily follow what exactly provided by mcore, is that right?

If we have single weight, then fallback is banned. If we have discrete weight plus that certain recipes don't have fully implemented grouped kernels, it will fallback to CPU based m_split and launch multistream gemms.

File: tests/unit_tests/transformer/moe/test_grouped_tensor_dispatcher_numerics.py (hunk 4 above)
Comment (zhongbozhu) [reply tests/unit_tests/transformer/moe/test_grouped_tensor_dispatcher_numerics.py:497]: sure, but I will gate it for this current PR

**Review by YangFei1990 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/extensions/transformer_engine.py [hunk 5]
```diff
@@ -2328,6 +2328,10 @@ def sharded_state_dict(
 
 if HAVE_TE and is_te_min_version("1.9.0.dev0"):
 
+    _TE_GROUPED_LINEAR_SUPPORTS_GROUPED_TENSOR = (
+        "use_grouped_tensor" in inspect.signature(te.pytorch.GroupedLinear.__init__).parameters
```
Comment (yashaswikarnati) [inline megatron/core/extensions/transformer_engine.py:2339]: just curious, if version check above is enough.. why would we need inspect signature as well here

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/experts.py:316]: I could not easily understand what this function is supposed to do based on the name.. just a suggestion, should we consider something like _add_router_weighted_packed_fc2_bias or smth along the lines

File: megatron/core/transformer/moe/experts.py [hunk 6]
```diff
@@ -634,9 +683,21 @@ def _fused_forward(
 
         # Apply padding if needed
         unpadded_tokens_per_expert = None
+        # Some dispatchers have already padded each expert's token segment before the tokens
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/experts.py:686]: just a nit; may be throughout the PR, the comments may be can be written little more concisely.

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (zhongbozhu) [reply megatron/core/transformer/moe/experts.py:316]: this naming is to keep parity with the old `_apply_bias`, and in general it's not bounded to `router_weighted_packed_fc2_bias` only

File: megatron/core/extensions/transformer_engine.py (hunk 5 above)
Comment (zhongbozhu) [reply megatron/core/extensions/transformer_engine.py:2339]: this feature should be enabled by the TE PR: https://github.com/NVIDIA/TransformerEngine/pull/3224 and before this PR, this `use_grouped_tensor` doesn't exist

in case the TE PR actually have some bugs (let's hope not), we might want guard it more in Mcore later

**Review by yashaswikarnati — APPROVED** [review]

**Review by hxbai — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]
